src/main/java/org/owasp/esapi/codecs/JSONCodec.java

```diff
@@ -0,0 +1,242 @@
+     * six-character sequences of \u0000 - \uffff. If the character is not encoded then a
+     * null character is returned.
+     *
+     * @param input a character sequence to decode
+     * @return the decoded version of the encoded character starting at index,
+     *     or null otherwise
+     *
+     * @throws IllegalArgumentException
+     *     if an invalid character sequence is encountered
+     */
+    public Integer decodeCharacter( PushbackSequence<Integer> input )
+        throws IllegalArgumentException {
+
+        input.mark();
+
+        Integer first = input.next(), second = null;
+        if ( first == null || first.intValue() != '\\' ) {
+            input.reset();
+            return null;
+        }
+
+        String errorMessage = null;
+
+        try
+        {
+            errorMessage = "Invalid JSON escape representation";
+
+            if ( (second = input.next()) == null ) {
+                throw new IllegalArgumentException();
+            }
+
+            // Per the RFC... Two-character sequence escape representations of some popular characters
+            switch ( second.intValue() ) {
+                case 'b': return (int)'\b';
+                case 'f': return (int)'\f';
+                case 'r': return (int)'\r';
+                case 'n': return (int)'\n';
+                case 't': return (int)'\t';
+                case '"': return (int)'\"';
+                case '/': return  (int)'/';
+                case '\\': return (int)'\\';
+            }
+
+            errorMessage = "Invalid JSON two-character escape representation";
+
+            // Per the RFC... All characters may be escaped as a six-character sequence: a reverse solidus,
+            // followed by the lowercase letter u, followed by four hexadecimal digits that encode the
+            // character's code point. The hexadecimal letters A through F can be uppercase or lowercase.
+            // So, for example, a string containing only a single reverse solidus character may be represented
+            // as "\u005C".
+            if ( second.intValue() == 'u' ) {
+
+                errorMessage = "Invalid JSON six-character escape representation";
+
+                return (convertToInt( input.next() ) << 12) +
+                       (convertToInt( input.next() ) <<  8) +
+                       (convertToInt( input.next() ) <<  4) +
+                       (convertToInt( input.next() ) <<  0);
+            }
+
+            // Do nothing. Fall into throw below.
+        }
+        catch (IllegalArgumentException e)
+        {
+            // Do nothing. Fall into throw below.
+        }
+
+        // Catch all. The escaped character sequence was invalid.
+        input.reset();
+        throw new IllegalArgumentException( errorMessage );
+    }
+
+    protected int charToCodepoint( Character ch ) {
+
+        final String s = Character.toString(ch);
+        assert (s.length() == 1) : "Ooops";
```

Comment:
Asserts are enabled by the Surefire plugin by default, so if this triggers we will see the output in the test dump.  Would you consider adding more context to the assert failure output to help identify this as the cause when inspecting from a terminal?

EG `"JSONCodec - Unable to convert specified character to codepoint.  Character String conversion must contain more than one character"`

Reply:
I don't feel strongly about this.  There is a lot of output in the tests, so the likelihood is that the stacktrace from any assertion failure will be more useful than the message itself.  This will get us in the right area, and I think most developers will be able to fill in the rest with the logic being applied.

Resolving this review thread.

Reply:
In https://github.com/ESAPI/esapi-java-legacy/pull/753#discussion_r1005491969, @jeremiahjstacey wrote:
> Asserts are enabled by the Surefire plugin by default, so if this triggers we will see the output in the test dump. Would you consider adding more context to the assert failure output to help identify this as the cause when inspecting from a terminal?

That still doesn't mean that we should rely on 'asserts'. The are not generally enabled in runtime by default and even if we are doing so in our Surefire plugin, I would feel better if anything that can be caused by a user accidentally providing incorrect input or not using the classes as per their Javadoc, should be an actual runtime check and not rely on `assert`. Rather, there should be an explicit check and if input is amok, then throw an `IllegalArgumentException`. In other words, do **NOT** use `assert` to check for preconditions. Using `asserts` for postconditions or invariants, I think is okay, as that generally means a program error on our part (assume we've validated all the inputs for sanity), but don't use them for precondition checking.

So, @noloader, if this is a precondition check using assertions, could you change it to an explicit check? Thanks. (I haven't checked, as I'm talking to my mother and trying to listen to her and type at the same time.)

Reply:
@kwwall per your comment on the assert statement I've unresolved that conversation.  Suggest waiting to merge pending updated resolution.

Reply:
See my previous comment here. If Unicode characters can extend from 0x0 to 0x10ffff (rather than 0xffff), is there any concern that this could result in a string whose length is >1?